README.md
# Barrin's Codex
[![License](https://img.shields.io/badge/License-MIT-blue)](https://opensource.org/licenses/MIT)
[![Python version](https://img.shields.io/badge/python-3.8-blue)](https://www.python.org/downloads/)
[![Validation](https://github.com/Spigushe/barrins-codex/actions/workflows/static.yml/badge.svg)](https://github.com/Spigushe/barrins-codex/actions/workflows/static.yml)
[![PyPI version](https://badge.fury.io/py/barrins-codex.svg)](https://badge.fury.io/py/barrins-codex)
[![Code Style](https://img.shields.io/badge/code%20style-black-black)](https://github.com/psf/black)
[![Imports: isort](https://img.shields.io/badge/%20imports-isort-%231674b1?style=flat&labelColor=ef8336)](https://pycqa.github.io/isort/)

A website about MtG Duel Commander Strategy

## Special Thanks
I started working with [lionel-panhaleux](https://github.com/lionel-panhaleux)
on a project during first lockdown (around April 2020). We worked on his
[Codex of the Damned](https://codex-of-the-damned.org/)
([repo](https://github.com/lionel-panhaleux/codex-of-the-damned)) to support
internationalisation. The whole structure of this project is derived from his
Codex, hence the name as a small tribute.

## Contributing
Contributions are welcome.
- [Pull Requests](https://github.com/Spigushe/barrins-codex/pulls) will be merged if they respect the general style.
- [Issues](https://github.com/Spigushe/barrins-codex/issues) will be dealt with as quickly as possible.

This site uses [Flask](https://flask.palletsprojects.com) to generate pages
dynamically .

## Installation
To install a working developpment version of the site, use `pip`:

```bash
python3 -m venv venv
source venv/bin/activate
pip install -e ".[dev]"
```

## Run the Codex
You can run the development version of the site using the `codex` entrypoint:

```bash
$ codex
* Serving Flask app "barrins_codex" (lazy loading)
* Environment: production
  WARNING: This is a development server. Do not use it in a production deployment.
  Use a production WSGI server instead.
* Debug mode: off
* Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
```

The Makefile has a command to start codex in debug : `make codex`

## Versioning
The version number take the form X.Y.Z where X, Y, and Z are non-negative
integers, and do not contain leading zeroes. X is the major version, Y is the
minor version, and Z is the patch version. Each element MUST increase
numerically. For instance: `1.9.0` -> `1.10.0` -> `1.11.0`.

Given a version number MAJOR.MINOR.PATCH, I increment the:
1. MAJOR version when I make a backend evolution or a design evolution,
1. MINOR version when I add content (most likely a match), and
1. PATCH version when I make bug and typo fixes.

Additional labels for pre-release and build metadata are available as
extensions to the MAJOR.MINOR.PATCH format.


CHANGELOG.rst
Changelog
=========

9.6.1 (unreleased)
------------------

- Nothing changed yet.


9.6.0 (2023-06-14)
------------------

- PARTENAIRES RÉCENTS : Clarification Star et colorité (suite)
- RESPONSIVE : Taille des images dans les bannières
- BANLIST : B&R du 29 mai 2023
- BACKEND : Change deprecated


9.5.3 (2023-01-01)
------------------

- URL : Suppression du ".html" en fin d'adresse
- URL : Utilisation des liens en ".html" sans rejet 404
- URL : N'affiche plus "index" par défaut sur l'accueil
- BANNIERE : Correction d'un comportement indésirable
- MATCHS : Retrait de la numérotation
- PARTENAIRES RÉCENTS : Clarification Star et colorité
- MATCHS : Tri par ordre alphabétique dans fiche pilote
- RESPONSIVE : Taille des images dans les articles


9.5.2 (2022-12-27)
------------------

- DECKS : Correction du positionnement des pilotes


9.5.1 (2022-12-27)
------------------

- STYLE : Modification des délais de transitions


9.5.0 (2022-12-27)
------------------

- STYLE : Ajout d'une transition pour réduire les bannières de match


9.4.2 (2022-12-27)
------------------

- INDEX : Affichage des matchs via un carousel
- INDEX : Ajout d'un lien dans le carousel
- YOSHIMARU : Clarification de la colorité
- MENU : Déplacement du style dans le fichier dédié
- STYLE : Refactoring en fichier JS
- SQUELETTE : Nettoyage du code et des fichiers inutiles
- STYLE : Ajout d'un constructeur de bannière
- ANAËL : Correction de l'image
- STYLE : Correction des classes de tip
- LEOVOLD : Ajout de la fiche du deck
- MATCH 37 : Ajout de la page de match


9.4.1 (2022-12-10)
------------------

- MATCH 36 : Ajout des liens dans les pages de deck
- APPARENCE : Amélioration des encarts de matchs sur la page d'accueil


9.4.0 (2022-12-10)
------------------

- KROXA : Ajout du deck
- CECILY-WERNOG : Ajout du deck
- MATCH 35 : Ajout du match
- RESPONSIVE : Correction du block SLD Universe Beyond
- GHYRSON : Ajout du deck
- NIV-MIZZET : Ajotu de la decklsit de NicolasP
- PILOTES : Ajout des liens et réordonnancement
- MATCH 36 : Ajout du match


9.3.0 (2022-11-27)
------------------

- Changement de pp pour nicolasp
- Ajout d'un nouveau pilote : Jules « Askeladden »
- Match 34 : Grist vs Raffine


9.2.1 (2022-11-14)
------------------

- Correction du contrôle de légalité


9.2.0 (2022-11-13)
------------------

- Ajout de la liste des généraux légaux


9.1.0 (2022-11-11)
------------------

- Retrait des lignes vides dans la génération
- Ajout de la mention "Sideboard" pour intégration Cockatrice
- Ajout d'un nouveau pilote : Nicolas « NicolasP »
- Création de la page du match #33
- Modification de l'affichage des decks dans les pages des pilotes
- Ajout de l'affichage des derniers matchs sur la page d'accueil
- Menu déroulant déclenché au passage de la souris
- Ajout d'un symbole pour expliciter le menu déroulant imbriqué


9.0.1 (2022-11-10)
------------------

- Correction du lien de génération


9.0.0 (2022-11-10)
------------------

- Ajout de la génération txt des decklists
- Ajout de la génération des boutons vers les listes
- Ajout de l'option "outline" pour les boutons
- Correction des templates de deck


8.3.1 (2022-11-02)
------------------

- GRENZO - Ajout d'un séparateur manquant
- INDEX - Nettoyage du code en trop de la page
- JURI - Ajout de la decktech


8.3.0 (2022-10-30)
------------------

- Correction de l'affichage des cartes bannies
- MODULE - Réduction du nombre d'instructions
- PILOTE - Ajout d'Alexandre « OneSauk »
- CONTEXTE - Ajout des symboles de mana
- MATCH 32 - Ajout de la page du match
- MATCH 32 - Ajout des decks
- MATCH 32 - MAJ des fiches pilotes
- GRENZO - Ajout du primer de Ejell de février 2021


8.2.3 (2022-10-09)
------------------

- Correction de la miniature de Tiamat
- Ajout des matches 31 à 40
- Ajout logo Github dans le footer
- Accessibility - Alt text de la bannière
- Accessibility - ID identiques dans le menu
- Accessibility - Liens sur logos trop petits
- Accessibility - First Painful Content géré en lazy loading
- SEO - Links are not crawlable
- Correction des espaces insécables pour les noms de cartes
- Ajout d'un système de sous-pages pour les archétypes
- Ajout d'un système de sous-pages pour les matchs


8.2.2 (2022-10-08)
------------------

- Correction da l'analyse du match 31


8.2.1 (2022-10-08)
------------------

- Correction du responsive de l'affichage des cartes
- Correction de l'affichage des cartes renommées après SLD


8.2.0 (2022-10-08)
------------------

- Refonte de l'affichage des cartes
- Match 31 : Raffine vs Soul of Windgrace


8.1.0 (2022-09-30)
------------------

- Match 30 : Gut vs Yoshimaru
- Correction du lien sur la page de Livio-Prava
- Ajout de la deck tech de Gitrog


8.0.0 (2022-09-29)
------------------

- Mise à jour de la définition de Midrange
- Ajout analyse 21. Aminatou vs Beamtown Bullies
- Ajout analyse 22. Raffine vs Tivit
- Image "firstprint" de chaque carte systématique
- Amélioration de la génération des liens Scryfall
- Ajout du lazy loading
- Fix images Tiamat
- Colorisation des noms de cartes et ajout d'espace insécable
- Correction grammaire analyse 22. Raffine vs Tivit
- Nouvelle catégorie : articles
- Fix définition Midrange
- Article sur la roue des archétypes
- Lien vers la roue dans match 22


7.2.0 (2022-09-24)
------------------

- Adding isort to project
- Changement de la miniature pour le deck Wilson + Background
- Match 29 : Erinis vs Raff
- Ajout deck tech Niv Mizzet Reborn


7.1.0 (2022-08-17)
------------------

- Refonte de l'apparence: Ajout de tag dans la navigation
- Ajout des pages dans le contexte de base
- Refonte de l'apparence: Réfection du menu
- Backend: Simplification des calls au helper
- Ajout des logos via Font Awesome Brand
- Ajout des img_crop dans le contexte de base
- Refonte de l'apparence: Decks index par type
- Ajout des définitions des macrotypes dans l'index
- Refonte de l'apparence: Match index par dizaine
- Refonte de l'apparence: Contexte passé au template 404

barrins_codex/__init__.py
import copy
import json
import os
import re
from itertools import islice
from urllib.parse import urlencode, urlparse

import flask
import jinja2.exceptions
import pkg_resources
import requests
import unidecode
from flask import Response, make_response, render_template, request
from markupsafe import Markup

from . import card_list, config, moxfield_decklist
from .navigation import HELPER

version = pkg_resources.Environment()["barrins-codex"][0].version
if version[-5:] == ".dev0":  # To pass tests on commits during development
    version = version[:-5]
    version = version[:-1] + str(int(version[-1]) - 1)

app = flask.Flask(__name__, template_folder="templates", static_folder="static")
app.jinja_env.trim_blocks = True
app.jinja_env.lstrip_blocks = True
config.configure_app(app)

if os.path.isfile("library.json"):
    with open("library.json", "r", encoding="utf-8") as file:
        CARDS = json.load(file)
else:
    CARDS = card_list.build()

BASE_CONTEXT = {
    "W": "⚪",
    "U": "🔵",
    "B": "⚫",
    "R": "🔴",
    "G": "🟢",
    "version": version,
    "cards": CARDS,
    "pilotes_habitue": [
        page for page in HELPER if HELPER.get(page, {}).get("cat") == "habitue"
    ],
    "pilotes_invite": [
        page for page in HELPER if HELPER.get(page, {}).get("cat") == "invite"
    ],
    "decks_agro": [
        page for page in HELPER if HELPER.get(page, {}).get("cat") == "agro"
    ],
    "decks_tempo": [
        page for page in HELPER if HELPER.get(page, {}).get("cat") == "tempo"
    ],
    "decks_controle": [
        page for page in HELPER if HELPER.get(page, {}).get("cat") == "controle"
    ],
    "decks_combo": [
        page for page in HELPER if HELPER.get(page, {}).get("cat") == "combo"
    ],
    "decks_midrange": [
        page for page in HELPER if HELPER.get(page, {}).get("cat") == "midrange"
    ],
    "decks_ban": [page for page in HELPER if HELPER.get(page, {}).get("cat") == "ban"],
    "needs_crop": [page for page in HELPER if HELPER.get(page, {}).get("crop")],
    "articles": [
        page for page in HELPER if HELPER.get(page, {}).get("cat") == "article"
    ],
    "last_matchs": list(
        islice(
            reversed(
                [page for page in HELPER if HELPER.get(page, {}).get("cat") == "match"]
            ),
            0,
            10,
        )
    ),
    "all_matchs": sorted(
        [page for page in HELPER if HELPER.get(page, {}).get("cat") == "match"]
    ),
    "num": ["1️⃣", "2️⃣", "3️⃣", "4️⃣", "5️⃣", "6️⃣", "7️⃣", "8️⃣", "9️⃣", "0️⃣"],
}


for card in CARDS:
    # Ajout des cartes en hover dans le contexte de base
    BASE_CONTEXT[card] = Markup(
        """<span class="card-name" scryfallId="{scryfallId}" data-tippy-content="
<div class='card-container'>
<img
    data-src='https://api.scryfall.com/cards/{scryfallId}?{query}'
    class='card-image'
/>
</div>">{name}</span>""".format(
            # replace spaces with non-breakable spaces in card names
            name=CARDS[card]["name"].replace(" ", "&#xA0;"),
            query="format=image&version=border_crop",
            scryfallId=CARDS[card]["id"],
        )
    )
    # Ajout des cartes img dans le contexte de base
    BASE_CONTEXT["img_" + card] = Markup(
        """
<img src="https://api.scryfall.com/cards/{scryfallId}?{query}"
    alt="{name}" class="col-md-3 float-md-end mx-md-1" max-width="100%"
    loading="lazy" />""".format(
            # replace spaces with non-breakable spaces in card names
            name=CARDS[card]["name"].replace(" ", " "),
            query="format=image&version=border_crop",
            scryfallId=CARDS[card]["id"],
        )
    )


def main():
    # print(HELPER)
    # print(BASE_CONTEXT)
    app.run()


# Defining Errors
@app.errorhandler(jinja2.exceptions.TemplateNotFound)
@app.errorhandler(404)
def page_not_found(error):
    context = copy.copy(BASE_CONTEXT)
    return flask.render_template("404.html", **context), 404


# Static file redirection
@app.route("/favicon.ico")
def favicon():
    return flask.redirect(flask.url_for("static", filename="img/favicon.ico"))


# Serve robots static file
@app.route("/robots.txt")
def static_from_root():
    return flask.send_from_directory(app.static_folder, flask.request.path[1:])


# Serve sitemap template
# code used from https://gist.github.com/Julian-Nash/aa3041b47183176ca9ff81c8382b655a
@app.route("/sitemap.xml")
def sitemap():
    host_components = urlparse(request.host_url)
    host_base = host_components.scheme + "://" + host_components.netloc

    urls = [host_base + p["self"].url for p in HELPER.values()]

    xml_sitemap = render_template("sitemap.xml", urls=urls, host_base=host_base)
    response = make_response(xml_sitemap)
    response.headers["Content-Type"] = "application/xml"

    return response


# Serve decklist routes
@app.route("/decklist/")
@app.route("/decklist/<moxfield_key>")
def download_list(moxfield_key=None):
    if not moxfield_key:
        return False

    response = moxfield_decklist.export(moxfield_key)

    return Response(
        response["list"],
        mimetype="text/plain",
        headers={
            "Content-Disposition": f"attachment;filename=\"{response['name']}.txt\""
        },
    )


# Serve commander informations
@app.route("/commanders")
def commanders():
    context = copy.copy(BASE_CONTEXT)
    commander = render_template("commanders_data.html", **context)

    return make_response(commander)


# Default route
@app.route("/")
@app.route("/<path:page>")
def index(page=None):
    redirect = False
    if not page:
        page = "index"
    if page[-5:] == ".html":
        page = page[:-5]
        redirect = True
    page = "/" + page
    if redirect:
        return flask.redirect(page, 301)

    context = copy.copy(BASE_CONTEXT)
    return render_template(page + ".html", **context)


def _build_url(page, _anchor=None, **params):
    url = page.url
    if params:
        url += "?" + urlencode(params)
    if _anchor:
        url += "#" + _anchor
    return url


def _link(page, name=None, _anchor=None, _class=None, **params):
    if not page or not page.url:
        return ""
    name = name or page.name
    url = _build_url(page, _anchor, **params)
    if _class:
        # _class = f"class='{_class} text-reset'"
        _class = f"class='{_class}'"
    else:
        _class = "class='text-reset'"
    return Markup(f'<a {_class} href="{url}">{name}</a>')


@app.context_processor
def linker():
    path = flask.request.path
    if path[-6:] == "/index":
        path = path[:-6]
    if path[-1:] == "/":
        path = path[:-1]

    def _url(page, _anchor=None, **params):
        return _build_url(HELPER.get(page, {}).get("self"), _anchor=_anchor, **params)

    def link(page, name=None, _anchor=None, **params):
        return _link(
            HELPER.get(page, {}).get("self"),
            name=name,
            _anchor=_anchor,
            **params,
        )

    def name(page, name=None):
        return name or HELPER.get(page, {}).get("self").name

    def top():
        return _link(HELPER.get(path, {}).get("top"), _class="text-reset")

    def next():
        return _link(HELPER.get(path, {}).get("next"), _class="next text-reset")

    def prev():
        return _link(HELPER.get(path, {}).get("prev"), _class="prev text-reset")

    def external(url, name, _class=None):
        if _class:
            _class = f"class='{_class}'"
        else:
            _class = "class='text-reset text-decoration-underline'"
        return Markup(
            f'<a {_class} target="_blank" rel="noreferrer" href="{url}">{name}</a>'
        )

    def title():
        try:
            if HELPER.get(path, {}).get("self").path != "":
                name = HELPER.get(path, {}).get("self").name
                return f"Barrin's Codex - {name}"
        except AttributeError:
            pass
        return "Barrin's Codex"

    def page_name():
        name = "Barrin's Codex"
        try:
            if HELPER.get(path, {}).get("self").path != "":
                name = HELPER.get(path, {}).get("self").name
        except AttributeError:
            pass
        return f"{name}"

    return dict(
        url=_url,
        link=link,
        title=title,
        name=name,
        page_name=page_name,
        top=top,
        next=next,
        prev=prev,
        external=external,
    )


@app.context_processor
def display_card():
    def _name(name):
        name = name
        name = unidecode.unidecode(name).lower()
        name = re.sub(r"[^a-zA-Z0-9]", "", name)
        return name

    def card_name_from_page(name):
        # Strange case where double-faced cards make the page reload
        # with `name` shorten by 5.
        page_name = (
            HELPER.get([key for key in HELPER if key.startswith(name)][0], {})
            .get("self")
            .name
        )
        if "❌" in page_name:
            page_name = page_name[2:]
        return page_name

    def img_crop(name, front=True):
        card = CARDS[_name(name)]
        query = "format=image&version=art_crop"
        if "faces" in card:
            if not front:
                query = query + "&face=back"
        return "https://api.scryfall.com/cards/" + f"{card['id']}?{query}"

    def img_card(name, front=True):
        if _name(name) not in CARDS.keys():
            fuzzy = re.sub(r"[^\w\s]", "", name)
            fuzzy = re.sub(r"\s+", "-", fuzzy)
            url = f"https://api.scryfall.com/cards/search?q={fuzzy}"
            r = requests.get(url)
            card = r.json()["data"][0]
            return card["image_uris"]["border_crop"]
        card = CARDS[_name(name)]
        query = "format=image&version=border_crop"
        if "faces" in card:
            if not front:
                query = query + "&face=back"
        return f"https://api.scryfall.com/cards/{card['id']}?{query}"

    def card_link(name, phrase=None):
        return Markup(
            """
<img src="https://api.scryfall.com/cards/{scryfallId}?{query}"
    alt="{name}" class="col-md-3 float-md-end mx-md-1" max-width="100%"
    loading="lazy" />""".format(
                # replace spaces with non-breakable spaces in card names
                name=phrase or CARDS[_name(name)]["name"].replace(" ", " "),
                query="format=image&version=border_crop",
                scryfallId=CARDS[_name(name)]["id"],
            )
        )

    def card_hover(name, phrase=None):
        return Markup(
            """<span class="card-name" scryfallId="{scryfallId}" data-tippy-content="
<div class='card-container'>
<img
    data-src='https://api.scryfall.com/cards/{scryfallId}?{query}'
    class='card-image'
/>
</div>">{name}</span>""".format(
                # replace spaces with non-breakable spaces in card names
                name=phrase or CARDS[_name(name)]["name"].replace(" ", " "),
                query="format=image&version=border_crop",
                scryfallId=CARDS[_name(name)]["id"],
            )
        )

    def card_crop(name, phrase=None, className=None):
        return Markup(
            """<img src="{crop}" alt="{name}" loading="lazy" {className} />""".format(
                # replace spaces with non-breakable spaces in card names
                name=phrase or CARDS[_name(name)]["name"].replace(" ", " "),
                crop=img_crop(name),
                className=f"class='{className}'" or "",
            )
        )

    def card_full(name, phrase=None, className=None):
        return Markup(
            """<img src="{full}" alt="{name}" loading="lazy" {className} />""".format(
                # replace spaces with non-breakable spaces in card names
                name=phrase or CARDS[_name(name)]["name"].replace(" ", " "),
                full=img_card(name),
                className=f"class='{className}'" or "",
            )
        )

    return dict(
        deck_name=card_name_from_page,
        img_crop=img_crop,
        img_card=img_card,
        card_link=card_link,
        card_hover=card_hover,
        card_crop=card_crop,
        card_full=card_full,
    )


@app.context_processor
def display_match():
    def match_name(page):
        try:
            if HELPER.get(page, {}).get("self").path != "":
                name = HELPER.get(page, {}).get("self").name
                return name
        except AttributeError:
            pass
        return "Barrin's Codex"

    return dict(match_name=match_name)


@app.context_processor
def players():
    def player_name(page):
        try:
            if HELPER.get(page, {}).get("self").path != "":
                name = HELPER.get(page, {}).get("self").name
                if name == "Apparitions":
                    return "Guest"
                return name.split(" ", 1)[0]
        except AttributeError:
            pass
        return "Barrin's Codex"

    def player_nickname(page):
        try:
            if HELPER.get(page, {}).get("self").path != "":
                name = HELPER.get(page, {}).get("self").name
                name = name.split(" ", 1)
                if len(name) == 1:
                    if name[0] == "Apparitions":
                        return "Pilote mystérieux·euse"
                    return name[0]
                return name[1][2:-2]
        except AttributeError:
            pass
        return "Barrin's Codex"

    return dict(player_name=player_name, player_nickname=player_nickname)


@app.context_processor
def decklist_processor():
    def decklist(url, name=None, outline=False):
        return Markup(
            """
<div class="col-12 d-flex flew-row mt-4 me-1">
    <a class="btn {btn}-secondary decklist col-10" role="button" target="_blank"
        href="{url}"><span title="Voir la decklist sur Moxfield">{name}</span>
    </a>
    <a class="btn {btn}-warning col-2 ms-1" role="button" href="{key}">
        <span title="Télécharger la decklist">
            <i class="fa-solid fa-download"></i>
        </span>
    </a>
</div>""".format(
                name=name or "Decklist",
                url=url,
                key="/decklist/" + url[31:],
                btn="btn-outline" if outline else "btn",
            )
        )

    return dict(decklist=decklist)
